src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/MediaS3BucketTenantEvents.cs

```diff
@@ -0,0 +1,134 @@
+        _amazonS3Client = amazonS3Client;
+        _options = options.Value;
+        S = localizer;
+        _logger = logger;
+    }
+
+    public override async Task ActivatingAsync()
+    {
+        if (!_options.CreateBucket ||
+            _shellSettings.State == Environment.Shell.Models.TenantState.Uninitialized ||
+            String.IsNullOrEmpty(_options.BucketName))
+        {
+            return;
+        }
+
+        _logger.LogDebug("Testing Amazon S3 Bucket {BucketName} existence", _options.BucketName);
+
+        try
+        {
+            var bucketExists = await AmazonS3Util.DoesS3BucketExistV2Async(_amazonS3Client, _options.BucketName);
+            if (bucketExists)
+            {
+                _logger.LogInformation("Amazon S3 Bucket {BucketName} already exists.", _options.BucketName);
+
+                return;
+            }
+
+            var bucketRequest = new PutBucketRequest
+            {
+                BucketName = _options.BucketName,
+                UseClientRegion = true
+            };
+
+            // Trying to create bucket.
+            var response = await _amazonS3Client.PutBucketAsync(bucketRequest);
+
+            if (!response.IsSuccessful())
+            {
+                _logger.LogError("Unable to create Amazon S3 Bucket {BucketName}", _options.BucketName);
+
+                return;
+            }
+
+            // Blocking public access for the newly created bucket.
+            var blockConfiguration = new PublicAccessBlockConfiguration
+            {
+                BlockPublicAcls = true,
+                BlockPublicPolicy = true,
+                IgnorePublicAcls = true,
+                RestrictPublicBuckets = true
+            };
+
+            await _amazonS3Client.PutPublicAccessBlockAsync(new PutPublicAccessBlockRequest
+            {
+                PublicAccessBlockConfiguration = blockConfiguration,
+                BucketName = _options.BucketName
+            });
+
+            _logger.LogDebug("Amazon S3 Bucket {BucketName} created.", _options.BucketName);
+        }
+        catch (AmazonS3Exception ex)
+        {
+            _logger.LogError(ex, "Unable to create Amazon S3 Bucket.");
+        }
+    }
+
+    public override async Task RemovingAsync(ShellRemovingContext context)
+    {
+        if (!_options.RemoveBucket || String.IsNullOrEmpty(_options.BucketName))
+        {
+            return;
+        }
+
+        try
+        {
+            var bucketExists = await AmazonS3Util.DoesS3BucketExistV2Async(_amazonS3Client, _options.BucketName);
+            if (!bucketExists)
+            {
+                return;
+            }
+
+            var bucketRequest = new DeleteBucketRequest
+            {
+                BucketName = _options.BucketName,
+                UseClientRegion = true
+            };
+
+            // Trying to delete bucket.
+            var response = await _amazonS3Client.DeleteBucketAsync(bucketRequest);
```

Comment:
Can't be done in a background task because we are deleting the tenant. By default we don't delete the bucket, this is a configurable option which by default is false.